server/datastore/mysql/software.go

```diff
@@ -415,30 +420,80 @@ func (ds *Datastore) applyChangesForNewSoftwareDB(
+func updateTargetedBundleIDs(ctx context.Context, tx sqlx.ExtContext, hostID uint, insertedSoftware []fleet.Software, updatedSoftware []softwareIDChecksum, bundleIDsToSoftware map[string]fleet.Software) error {
 	if len(bundleIDsToSoftware) == 0 {
 		return nil
 	}
 
-	updateSoftwareStmt := `UPDATE software SET software.name = ?, software.name_source = 'bundle_4.67' WHERE software.bundle_identifier = ?`
-
-	hostSoftwareStmt := `
-		INSERT IGNORE INTO host_software
-			(host_id, software_id, last_opened_at)
-		VALUES
-			(?, (SELECT id FROM software WHERE bundle_identifier = ? AND name_source = 'bundle_4.67' ORDER BY id DESC LIMIT 1), ?)`
+	// inserted/updated software id -> software whose name has changed (name / last_opened_at)
+	softwareIDToUpdate := make(map[uint]fleet.Software)
+	var softwareIDs []uint
 
-	for k, v := range bundleIDsToSoftware {
-		if _, err := tx.ExecContext(ctx, updateSoftwareStmt, v.Name, k); err != nil {
-			return ctxerr.Wrap(ctx, err, "update software names")
+	// checking inserted software
+	for _, software := range insertedSoftware {
+		if software.BundleIdentifier != "" {
+			if updatedSoftware, needsUpdate := bundleIDsToSoftware[software.BundleIdentifier]; needsUpdate {
+				softwareIDToUpdate[software.ID] = updatedSoftware
+				softwareIDs = append(softwareIDs, software.ID)
+			}
 		}
+	}
 
-		if _, err := tx.ExecContext(ctx, hostSoftwareStmt, hostID, v.ID, v.LastOpenedAt); err != nil {
-			return ctxerr.Wrap(ctx, err, "insert host software")
+	// checking existing software
+	for _, software := range updatedSoftware {
+		if software.BundleIdentifier != nil && *software.BundleIdentifier != "" {
+			if updatedSoftware, needsUpdate := bundleIDsToSoftware[*software.BundleIdentifier]; needsUpdate {
+				softwareIDToUpdate[software.ID] = updatedSoftware
+				softwareIDs = append(softwareIDs, software.ID)
+			}
 		}
 	}
-	return nil
+
+	if len(softwareIDToUpdate) == 0 {
+		return nil
+	}
+
+	const batchSize = 100
+	return common_mysql.BatchProcessSimple(softwareIDs, batchSize, func(batch []uint) error {
+		updateArgs := make([]interface{}, 0, len(batch)*3)
+		updateCases := make([]string, 0, len(batch))
+		updateIDs := make([]string, 0, len(batch))
+
+		insertArgs := make([]interface{}, 0, len(batch)*3)
+		insertValues := make([]string, 0, len(batch))
+
+		for _, id := range batch {
+			software := softwareIDToUpdate[id]
+
+			updateCases = append(updateCases, "WHEN ? THEN ?")
+			updateArgs = append(updateArgs, id, software.Name, id)
+			updateIDs = append(updateIDs, "?")
+
+			insertValues = append(insertValues, "(?, ?, ?)")
+			insertArgs = append(insertArgs, hostID, id, software.LastOpenedAt)
+		}
+
+		updateStmt := fmt.Sprintf(
+			`UPDATE software SET name = CASE id %s END, name_source = 'bundle_4.67' WHERE id IN (%s)`,
+			strings.Join(updateCases, " "),
+			strings.Join(updateIDs, ","),
+		)
+
+		if _, err := tx.ExecContext(ctx, updateStmt, updateArgs...); err != nil {
```

Comment:
IIRC, one of the core issues last time was that this updating logic was executing inside an already very busy transaction. This lead to deadlocks, which lead to the DB issues.

I think we should do this update in a separate transaction.